backend/kafka_backend/kafka_backend.go

```diff
@@ -0,0 +1,135 @@
+
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/skroutz/downloader/job"
+)
+
+var eventsWG sync.WaitGroup
+
+// KafkaBackend notifies about a job completion by producing to a Kafka topic.
+type KafkaBackend struct {
+	producer *kafka.Producer
+	reports  chan job.CallbackInfo
+}
+
+// ID returns "kafka".
+func (b *KafkaBackend) ID() string {
+	return "kafka"
+}
+
+// Start sets up a kafka producer and a channel for delivery reports.
+func (b *KafkaBackend) Start(ctx context.Context, cfg map[string]interface{}) error {
+	var err error
+
+	kafkaCfg := make(kafka.ConfigMap)
+
+	for k, v := range cfg {
+		switch v.(type) {
+		case float64:
+			err := kafkaCfg.SetKey(k, int(v.(float64)))
+			if err != nil {
+				return err
+			}
+		default:
+			err := kafkaCfg.SetKey(k, v)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	b.producer, err = kafka.NewProducer(&kafkaCfg)
+	if err != nil {
+		return err
+	}
+
+	b.reports = make(chan job.CallbackInfo)
+
+	// start a go routine to monitor Kafka's Events channel
+	eventsWG.Add(1)
+	go b.transformStream(ctx, &eventsWG)
+
+	return nil
+}
+
+// Notify produces (asynchronously) a payload message into a kafka topic using
+// the producer. Kafka will send the delivery report to the default Events delivery channel.
+func (b *KafkaBackend) Notify(destination string, cbInfo job.CallbackInfo) error {
+	if cbInfo == (job.CallbackInfo{}) {
+		return errors.New("Received zero value composite literal for callback info")
+	}
+
+	payload, err := cbInfo.Bytes()
+	if err != nil {
+		return err
+	}
+
+	message := &kafka.Message{
+		TopicPartition: kafka.TopicPartition{Topic: &destination, Partition: kafka.PartitionAny},
+		Value:          payload,
+	}
+	err = b.producer.Produce(message, nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// DeliveryReports returns a channel of emmited callback info events
+func (b *KafkaBackend) DeliveryReports() <-chan job.CallbackInfo {
+	return b.reports
+}
+
+// Stop closes the reports channel and the producer.
+// It also handles gracefully the termination of any go routines
+// that the kafka backend started.
+func (b *KafkaBackend) Stop() error {
+	close(b.reports)
+	eventsWG.Wait()
+	b.producer.Close()
+	return nil
+}
+
+// transformStream iterates over the Events channel of Kafka, transforms
+// each message to a callback info object and enqueues the callback info object
+// to our reports channel.
+func (b *KafkaBackend) transformStream(ctx context.Context, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case e, ok := <-b.producer.Events():
+			if !ok {
+				return
+			}
+
+			switch ev := e.(type) {
+			case *kafka.Message:
+				var cbInfo job.CallbackInfo
+
+				err := json.Unmarshal(ev.Value, &cbInfo)
+				if err != nil {
+				} else {
+					if ev.TopicPartition.Error != nil {
+						fmt.Printf("Delivery failed: %v\n", ev.TopicPartition)
+						cbInfo.Error = ev.TopicPartition.Error.Error()
```

Comment:
Set `CallbackInfo.Success = false`

Reply:
@agis on a second thought, IMHO this seems semantically incorrect and it probably needs discussion. `CallbackInfo.Success` is set according to if a `Job.DownloadState` was successful or not, but in this case we alter the `CallbackInfo.Success` field to indicate the success or failure of communicating a Job's result. In simpler words, we map two completely different and unrelated notions to a single field. Maybe we should create two fields `CallbackInfo.Delivered (bool)` and `CallbackInfo.DeliveryError (string)` and use these appropriately wherever they are needed. What do you think?